### PR TITLE
Fix 'arguments' check in class field initializer or static initialization block

### LINKED ISSUE #44269: Spec: 'arguments' is not allowed in class field initializer or static initialization block
### 🔎 Search Terms

'arguments' is not allowed in class field initializer or static initialization block

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABATQBQEpEG8CwAoRQxAJwFMoRikIAbAQwGcHEAVbfIzuxAXkTuIBzEAFtSYKAw5EAvvjl4gA)

### 💻 Code

```ts
function Y() {
    return class T {
       a = arguments
    }
}
```

### 🙁 Actual behavior

No error

### 🙂 Expected behavior

TS Error: 'arguments' is not allowed in class field initializer or static initialization block

image

https://tc39.es/ecma262/multipage/ecmascript-language-functions-and-classes.html#sec-class-definitions-static-semantics-early-errors

```
FieldDefinition :
    ClassElementName Initializer[opt]
        It is a Syntax Error if Initializer is present and ContainsArguments of Initializer is true.
```

### PR DESCRIPTION
**Pull request by exoticknight**
Fixes #44269

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -24223,6 +24223,11 @@ namespace ts {
                     }
                 }
 
+                const firstFunctionLikeContainerOrClassLikeContainer = findAncestor(node.parent, node => isClassLike(node) || isFunctionLike(node))!;
+                if (firstFunctionLikeContainerOrClassLikeContainer.kind === SyntaxKind.ClassExpression || firstFunctionLikeContainerOrClassLikeContainer.kind === SyntaxKind.ClassDeclaration) {
+                    error(node, Diagnostics.arguments_is_not_allowed_in_class_field_initializer_or_static_initialization_block);
```
Comment (Jack-Works) [inline src/compiler/checker.ts:24228]: And it should not execute the following two lines.

You should add

```js
return errorType;
```

Comment (exoticknight) [reply src/compiler/checker.ts:24228]: LGTM

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -24223,6 +24223,12 @@ namespace ts {
                     }
                 }
 
+                const firstFunctionLikeContainerOrClassLikeContainer = findAncestor(node.parent, node => isClassLike(node) || isFunctionLike(node))!;
```
Comment (Jack-Works) [inline src/compiler/checker.ts:24226]: This is not correct. Consider this code:

```js
function a() {
    return (new class { a() { return arguments } }).a()
}
```

It's valid in JavaScript. Please make sure you check the arguments is in the banned position of a class member

Comment (DanielRosenwasser) [inline src/compiler/checker.ts:24226]: Consider moving this higher and reusing it for `container`

```ts
const firstFunctionLikeContainerOrClassLikeContainer = findAncestor(node.parent, node => isClassLike(node) || isFunctionLike(node))!;
const container = isClassLike(firstFunctionLikeContainerOrClassLikeContainer) ?
    findAncestor(firstFunctionLikeContainerOrClassLikeContainer.parent, isFunctionLike)! :
    firstFunctionLikeContainerOrClassLikeContainer;
```

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -24223,6 +24223,12 @@ namespace ts {
                     }
                 }
 
+                const firstFunctionLikeContainerOrClassLikeContainer = findAncestor(node.parent, node => isClassLike(node) || isFunctionLike(node))!;
+                if (firstFunctionLikeContainerOrClassLikeContainer.kind === SyntaxKind.ClassExpression || firstFunctionLikeContainerOrClassLikeContainer.kind === SyntaxKind.ClassDeclaration) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:24227]: isClassLike

**Review by DanielRosenwasser — COMMENTED** [review]
Technically there's an `isInPropertyInitializer` but the suggestions I've given seem cheaper.

File: src/compiler/diagnosticMessages.json [hunk 4]
```diff
@@ -3336,6 +3336,10 @@
         "category": "Error",
         "code": 2812
     },
+    "'arguments' is not allowed in class field initializer or static initialization block.": {
```
[the remark this answers is not in the document]
Comment (Jack-Works) [reply src/compiler/diagnosticMessages.json:3339]: Oh, the error message is copied from Chrome. But Chrome does not support static initialization block either, that's strange.

**Comment by exoticknight** [comment]
@DanielRosenwasser `isInPropertyInitializer` works well, thx for your review!

**Comment by exoticknight** [comment]
only node14 failed，werid

**Review by sandersn — APPROVED** [review]
Looks good to me -- what do you think @DanielRosenwasser ?

@exoticknight you can merge now from main to get diagnosticMessages up to date, or you can wait until @DanielRosenwasser signs off, it's up to you.

**Comment by sandersn** [comment]
I looked over the comment history and I think this is ready to go once CI finishes one more time.

**Pull request merged**